PPM Calibrated Energy/Background Calibration

KaushikMalapati opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
## Description
<!--- Describe your changes in detail -->
New way of calculating pulse-by-pulse energy using responsivity calibrated for each power meter, and a pv trigger to collected background voltage. I also added the Tc2_System library to get the cycle time. 

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
https://jira.slac.stanford.edu/browse/ECS-875

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Not yet tested. 

## Where Has This Been Documented?
<!--  Include where the changes made have been documented. -->
<!--  This can simply be  a comment in the code or updating a docstring -->
Code is commented.

## Pre-merge checklist
- [ ] Code works interactively
- [ ] Code contains descriptive comments
- [ ] Test suite passes locally
- [ ] Libraries are set to ``Always Newest`` version (``Library, *``)
- [ ] Committed with ``pre-commit`` or ran ``pre-commit run --all-files``

ZLLentz commented on lcls-twincat-common-components/Library/Devices/PPM/FB_PPM_PowerMeter.TcPOU:
```diff
@@ -77,10 +85,71 @@ VAR
     '}
     fCalibMJRatio: LREAL;
```
Feel free to strip out all of the old code that does similar things but wrong

ZLLentz commented on lcls-twincat-common-components/Library/Devices/PPM/FB_PPM_PowerMeter.TcPOU:
```diff
@@ -77,10 +85,71 @@ VAR
+    // Linked to beam rate
+    {attribute 'pytmc' := '
+        pv: BeamRateFromEpics
+        link: TPG:SYS0:1:DST04:RATE
```
Is there some variability here depending on which accelerator is in use? Do we need to have some special handling for this?

ZLLentz commented on lcls-twincat-common-components/Library/Devices/PPM/FB_PPM_PowerMeter.TcPOU:
```diff
@@ -103,6 +172,26 @@ fbGetPMVoltage(
+    fBackground := fBackgroundCollectionSum / fBackgroundCollectionCyclesDone;
+END_IF
+// Getting wattage and energy
+IF fResponsivity <> 0 AND fFrequency <> 0 THEN
+    fResponWattage := (fVoltage - fBackground) / fResponsivity;
+    fResponMJ := fResponWattage / fFrequency;
+END_IF
```
I think we should still calculate the Wattage even if the frequency is zero, because it's possible that our frequency information is actually wrong.

Maybe we should set the MJ to some obviously invalid value in these cases?

ZLLentz commented on lcls-twincat-common-components/Library/Devices/PPM/FB_PPM_PowerMeter.TcPOU:
Overall I think this is a solid implementation of the request :+1: does the job seemingly correctly and robustly

ZLLentz commented on lcls-twincat-common-components/Library/Devices/PPM/FB_PPM_PowerMeter.TcPOU:
```diff
@@ -77,10 +85,71 @@ VAR
+    // Number of cycles to collect background voltages for
+    fBackgroundCollectionCycles: UDINT;
+    // Number of cycles that background voltages have already been collected
+    fBackgroundCollectionCyclesDone: UDINT;
```
f is the prefix for floating points, but this is actually assigned as a UDINT and other fs are assigned as BOOL

ZLLentz commented on lcls-twincat-common-components/Library/Devices/PPM/FB_PPM_PowerMeter.TcPOU:
```diff
@@ -103,6 +172,26 @@ fbGetPMVoltage(
+// (Re)starting backround collection, will gracefully restart
+IF fBackgroundCollect = TRUE THEN
+    fBackgroundCollect := FALSE;
+    fBackgroundCollectionCycles := 10_000_000 * fBackgroundCollectionTime / _TaskInfo[GETCURTASKINDEXEX()].CycleTime;
```
Any compiler warnings here for the implicit floating point -> int conversion? We can make it explicit to get rid of these via the `LREAL_TO_UDINT` function.

ZLLentz commented:
The last thing to consider here is if there are any unit tests we can possibly make for this. My intuition is that we can make one for the timer and the calc but that they would have dubious value.

ZLLentz commented on lcls-twincat-common-components/Library/Devices/PPM/FB_PPM_PowerMeter.TcPOU:
```diff
@@ -6,6 +6,9 @@ VAR_INPUT
     bVetoTempFFO: BOOL;
     sDeviceName: STRING;
+    eEnumGet: E_PPM_States;
+    stYAxisState: DWORD;
```
Review note to help myself: axis state dword is implicitly an enum as described on this page: https://infosys.beckhoff.com/english.php?content=../content/1033/tcplclib_tc2_mc2/70133899.html&id=

ZLLentz commented on lcls-twincat-common-components/Library/Devices/PPM/FB_PPM_PowerMeter.TcPOU:
```diff
@@ -35,123 +39,111 @@ VAR
-    fbResponMJBuffer: FB_LREALBuffer;
+
+    FFO: FB_FastFault :=(
+        i_Desc := 'Fault occurs when the temprature trip point is reached',
```
```suggestion
        i_Desc := 'Fault occurs when the temperature trip point is reached',
```

ZLLentz replied:
Aside from the typo, maybe this description should be more clear that this is about the per-pulse MJ reading being too high for the powermeter? I'm not 100% sure how to phrase this, and it technically is a temperature, but I suspect that calling it a temperature will confuse people who are looking at the other thermocouples on that device that give us readouts in Celcius.

KaushikMalapati replied:
How about "Fault occurs when the per-pulse energy reading is high enough to damage the power meter?" Also, do you know of a better typecode for this? I meant to use 16#400, not 16#f400, which is the optics code based on https://confluence.slac.stanford.edu/display/L2SI/PMPS+Type+Codes.

ZLLentz replied:
I think just that's a good description.
For the typecode: I think you should pick a new number close to a similar fault and add it to the confluence table. I think this is functionally distinct enough from the other codes that it deserves a new number.

KaushikMalapati replied:
Added code 16#500 - Pulse energy measured by PPM is high enough to damage power meter.

ZLLentz commented on lcls-twincat-common-components/Library/Devices/PPM/FB_PPM_PowerMeter.TcPOU:
```diff
@@ -99,19 +160,91 @@ fbGetPMVoltage(
+END_IF
+
+CASE PMPS_GVL.cstFullBeam.neV OF
+    0..300: 				bOverAllowableEnergy := fCalibMJ > 2;
```
Low priority nitpick: something strange is going on with the tab spacing here

KaushikMalapati replied:
![image](https://github.com/pcdshub/lcls-twincat-common-components/assets/80156796/341e734f-7514-4e3e-876e-85c8c9089924)
I tabbed the first case so the body would line up with the rest

ZLLentz commented on lcls-twincat-common-components/Library/Devices/PPM/FB_PPM_PowerMeter.TcPOU:
```diff
@@ -6,6 +6,18 @@ VAR_INPUT
+        field: EGU V/W
+    '}
+    fResponsivity: LREAL;
+    fVoltageCeiling: LREAL;
```
This input appears to be unused

ZLLentz commented on lcls-twincat-common-components/Library/Devices/PPM/FB_PPM_PowerMeter.TcPOU:
```diff
@@ -28,61 +40,110 @@ VAR
 END_VAR
-]]></Declaration>
+VAR CONSTANT
+    fWavelengthEdge1: REAL := 300 + 1.0E-44;
```
Is this 1.0E-44 related to the smallest increment of a REAL?
Is there a reason these can't be integers?

KaushikMalapati replied:
It is and there is not.

ZLLentz commented on lcls-twincat-common-components/Library/Devices/PPM/FB_PPM_PowerMeter.TcPOU:
```diff
@@ -99,19 +160,91 @@ fbGetPMVoltage(
+CASE PMPS_GVL.cstFullBeam.neV OF
+    0..300: 				bOverAllowableEnergy := fCalibMJ > 2;
+    fWavelengthEdge1..1000: bOverAllowableEnergy := fCalibMJ > 4;
+    fWavelengthEdge2..1600: bOverAllowableEnergy := fCalibMJ > 0.5;
```
The lower edges here are defined as constants above, but the upper edges are hard-coded here. There's probably a way to implement this more consistently.
Something like (details/syntax issues notwithstanding)
```
0..fWaveLengthEdge1
fWavelengthEdge1+1...fWavelengthEdge2
```

ZLLentz commented on lcls-twincat-common-components/Library/Devices/PPM/FB_PPM_PowerMeter.TcPOU:
```diff
@@ -99,19 +160,91 @@ fbGetPMVoltage(
-fbCalibBaseBuffer(
-    bExecute := TRUE,
-    fInput := fCalibBase,
-    arrOutput => fCalibBaseBuffer);
```
Is it worthwhile to replace this with a wattage buffer?

ZLLentz commented on lcls-twincat-common-components/Library/Devices/PPM/FB_PPM_PowerMeter.TcPOU:
```diff
@@ -99,19 +160,91 @@ fbGetPMVoltage(
+END_CASE
+
+// Getting wattage and energy
+fPulseWattage := (fVoltage - fBackgroundVoltage) / fResponsivity;
```
`fResponsivity` can cause a zero div here

ZLLentz reviewed (commented):
Partial review- still need to look at the background calculation when I get back later
Overall I like this and I'm mostly nitpicking

ZLLentz commented on lcls-twincat-common-components/Library/Devices/PPM/FB_PPM_PowerMeter.TcPOU:
```diff
@@ -99,19 +160,91 @@ fbGetPMVoltage(
+    fBackgroundVoltageSum := 0;
+    tonBackgroundAutoCollecting(IN := FALSE);
+    udBackgroundVoltageBufferIndex := 1;
+ELSIF eEnumGet = E_PPM_States.OUT AND stYAxisState = 0 THEN
```
Following along: if we didn't just start a move and we're standing still in the OUT state, wait 1 second and then sum the accumulated background voltages?

KaushikMalapati replied:
Yes. I'm using fBackgroundVoltageAcc so to accumulate every cycle so that the array is a reasonable size and to make dumping values when a move is started simpler.

ZLLentz commented on lcls-twincat-common-components/Library/Devices/PPM/FB_PPM_PowerMeter.TcPOU:
```diff
@@ -28,61 +40,110 @@ VAR
         io: io
     '}
-    fCalibRelRatio: LREAL;
+    BACKGROUND_MODE: (Manual, Passive) := Passive;
```
Is this some sort of inline enum definition? I hadn't seen this before

KaushikMalapati replied:
I think I found it here and decided to use it since it was pretty specific. https://alltwincat.com/2021/11/16/local-enumerations/
I could make a normal DUT type if that'd be clearer.

ZLLentz replied:
Nah, keep it in, this is great and I'm happy you've shown me how to use it.

ZLLentz commented on lcls-twincat-common-components/Library/Devices/PPM/FB_PPM_PowerMeter.TcPOU:
```diff
@@ -99,19 +160,91 @@ fbGetPMVoltage(
+        tonBackgroundAutoCollecting( IN := FALSE);
+        fBackgroundVoltageStale := fBackgroundVoltageBuffer[udBackgroundVoltageBufferIndex];
+        fBackgroundVoltageBuffer[udBackgroundVoltageBufferIndex] := fBackgroundVoltageAcc;
+        IF udBackgroundVoltageBufferIndex > 100 THEN
+            fBackgroundVoltageSum := fBackgroundVoltageSum - fBackgroundVoltageStale;
+        END_IF
+        fBackgroundVoltageSum := fBackgroundVoltageSum + fBackgroundVoltageAcc;
```
Is it possible for these voltage sums to accumulate floating point addition errors if we keep subtracting from and adding to them like this?

KaushikMalapati replied:
There will be some error, but I don't expect it to be large enough to be significant. The alternative I know of is to sume over the array every cycle (or every n cycles), but it would make those cycles significantly longer.

ZLLentz replied:
You're probably right. We should give this a try and see if there's any other nuances.

KaushikMalapati replied:
Wanted to briefly revisit this. Let's say we were worried about floating-point error accumulation, and decided to sum over a hundred-element array every second. How would you quantify the increase in cycle time this creates for the cycle where that sum occurs and the likelihood of exceeding the cycle time for a program that has ppm function blocks?

ZLLentz reviewed (commented):
Finishing the review of the background calculator- questions and nitpicks
And now I'll go back to respond to your comments

ZLLentz commented:
Reading back through my comments this is looking pretty mature- we should test soon in TMO

KaushikMalapati commented on lcls-twincat-common-components/Library/Devices/PPM/FB_PPM_PowerMeter.TcPOU:
```diff
@@ -99,19 +160,91 @@ fbGetPMVoltage(
-// Power meter calibration
-fCalibBase := (fVoltage + fCalibRelOffset) * fCalibRelRatio;
-fCalibMJ := fCalibBase * fCalibMJRatio;
+IF R_TRIG(CLK:=(eEnumGet <> E_PPM_States.OUT OR stYAxisState <> 0)) THEN
```
Thanks for catching the R_TRIG misuse. The logic is correct. When we are no longer stationary and at the OUT state, dump five seconds worth of background voltages, and set background variables to whatever they are expected when background measurements restart. This way it will automatically stop and start, with background collected even if in manual mode. Also, by resetting udBackgroundVoltageBufferIndex, we don't have to clear the array

KaushikMalapati commented:
I added the original PPM and PPM_PowerMeter fbs back so that we can test one at a time. I also responded to all the feedback you've given up to now, added some more comments in the code, and added a manual reset for the auto background reset collection. I considered removing the manual mode collection trigger, but decided to keep it so that a user can get a background voltage over a specified time period.

ZLLentz commented:
@KaushikMalapati did the tests of this go well? I'm looking to deploy #68 to kfe/tmo/rix in early august so I want to see if this could be bundled with that update.

KaushikMalapati commented:
> @KaushikMalapati did the tests of this go well? I'm looking to deploy #68 to kfe/tmo/rix in early august so I want to see if this could be bundled with that update.

I ran into an odd bug when trying to run tmo-motion using one of the new function blocks. I got logging errors which went away when I commented out PRG_4_LOG, but then while the program was running no variables were being set to their initial values and stayed at their defaults, even when I tried to force write. Not sure where the problem is, but Tong said she could help me with this after she and Nick finish their update.

KaushikMalapati commented:
I tested this on IM4K2, although without any beam rate and not the fast fault. I believe that everything works now, but I found 

- a few places where I was indexing out of bounds by not modding my index variable by 100 for the array I'm treating as circular, 
- using maximums where I should have used minimums when calculating the background voltage, 
- using wrong units in several places
- not dividing fBackgroundVoltageAcc by the number of measurements it had been summed over

I also changed the manual mode so that you can enter it without automatically starting to collect values. This would be useful if you already have a manual value that you want to reuse or if you want to enter manual mode, then set the collection time, then start collection. 

I reopened a comment about whether both accelerators use the same beam rate because looking back on it, I am not sure. 

Everything looks reasonable to me right now, but considering how much I had to change I would appreciate a once-over in case I missed anything else.

ZLLentz commented on lcls-twincat-common-components/Library/Library.plcproj:
```diff
@@ -15,7 +15,7 @@
     <LibraryReferences>{78c19189-384c-4a07-a84d-94cfbb123508}</LibraryReferences>
     <CombineIds>false</CombineIds>
-    <Title>lcls-twincat-common-components</Title>
+    <Title>lcls-twincat-cc-km</Title>
```
Reminder to switch back to the normal library name before merge

KaushikMalapati replied:
Switched back

ZLLentz commented on lcls-twincat-common-components/Library/Devices/PPM/FB_PPM_R.TcPOU:
Is this new FB representing new hardware that necessitates a completely separate function block? My understanding is that this was a revision/upgrade of behavior that may have been better suited to updating the other function block in-place.

KaushikMalapati replied:
It doesn't need new hardware, I got rid of the new _R function blocks

ZLLentz commented on lcls-twincat-common-components/Library/Devices/PPM/FB_PPM_R_PowerMeter.TcPOU:
```diff
@@ -0,0 +1,306 @@
+    eEnumGet: E_PPM_States;
+    stYAxisState: DWORD;
+
+    // Calibrated for each gentec, should be passed as input or set via pv
```
I think values set via PV will always be overridden cyclically by the VAR_INPUT variable in prod, but perhaps this should be a PLC-side parameter anyway?

KaushikMalapati replied:
I made it input only so it can be displayed on the lucid screen

ZLLentz commented on lcls-twincat-common-components/Library/Devices/PPM/BEAMLINE.TcDUT:
This is a clean and explicit way to specify which beamline we're on, but also we do this a different way in the PMPS lib using compiler defines that's worth looking at:

https://github.com/pcdshub/lcls-twincat-pmps/blob/8dba453befd302b8dd2d2e4236c991742574cd25/lcls-twincat-pmps/PMPS/HelperFunctions/PE_Ranges.TcPOU#L26-L30

Summary: each PLC is defining either "K" (soft) or "L" (hard) in their compiler defines and the PMPS lib is setting itself up differently depending on which one is chosen.

ZLLentz reviewed (commented):
Some minor comments and questions
This diff is actually pretty hard to review because it's an edited copy of the original FB

ZLLentz commented on lcls-twincat-common-components/Library/Devices/PPM/FB_PPM_PowerMeter.TcPOU:
```diff
@@ -28,61 +41,145 @@ VAR
+    '}
+    uAccelerator: UINT;
+    {attribute 'pytmc' := '
+        pv: NCBeamRateFromEpics
+        link: TPG:SYS0:1:DST03:RATE
+    '}
+    uSC_HXR_Frequency: UINT;
+    {attribute 'pytmc' := '
+        pv: SCBeamRateFromEpics
+        link: TPG:SYS0:1:DST04:RATE
     '}
-    fCalibBaseBuffer: ARRAY[1..1000] OF LREAL;
+    uSC_SXR_Frequency: UINT;
+    // Placeholders until I find the rights pvs for normal conducting
+        {attribute 'pytmc' := '
+        pv: NCBeamRateFromEpics
+        link: TPG:SYS0:1:DST03:RATE
+    '}
+    uNC_HXR_Frequency: UINT;
+    {attribute 'pytmc' := '
+        pv: SCBeamRateFromEpics
+        link: TPG:SYS0:1:DST04:RATE
+    '}
+    uNC_SXR_Frequency: UINT;
```
I think the `pv: ` lines in this block might have some typos/copy-paste errors since some of them are the same

ZLLentz commented on lcls-twincat-common-components/Library/Devices/PPM/FB_PPM_PowerMeter.TcPOU:
```diff
@@ -28,61 +41,145 @@ VAR
+    // Used to get beam mode and rate
     {attribute 'pytmc' := '
-        pv: CALIB_BUFFER
-        io: input
+        pv: AccTypeFromEpics
+        link: TPR:FEES:GD02:0:MODE.RVAL
+    '}
+    uAccelerator: UINT;
```
This link for `uAccelerator` is a bit suspicious, it has `FEES` in it which implies it might only be correct for the soft-xray line and there may be an alternate PV for the accelerator type on the hard x-ray line

ZLLentz reviewed (commented):
Everything looks good here except we need to lock down these mysterious PVs that nobody seems to know anything about

ZLLentz commented on lcls-twincat-common-components/Library/Devices/PPM/FB_PPM_PowerMeter.TcPOU:
```diff
@@ -28,61 +41,151 @@ VAR
+        pv: SXR_NC_BeamRateFromEpics
+        link: EVNT:SYS0:1:NC_SOFTRATE
+    '}
+    uSXR_NC_Frequency: UINT;
```
Small data typing thing, some of these links are DBR_DOUBLE in EPICS.
Do you know what EPICS does when the types don't match?
Hopefully it does something sensible like rounding.

KaushikMalapati replied:
I assume it does round, but I changed the two EVNT:SYS0:1 links to be reals.

ZLLentz reviewed (dismissed):
I think everything lines up now and the real PVs look sensible and if they came from Ryan we should be good.

I think the EPICS types thing probably works as we expect

ZLLentz commented:
@KaushikMalapati is it time to release and deploy a tag for this repo?
